plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -53,123 +69,124 @@ in
--}
-type Arity = [TermOrType]
-
--- | Is the next argument a term or a type?
-data TermOrType =
-    MkTerm | MkType
-    deriving stock (Eq, Show)
-
-instance Pretty TermOrType where
-  pretty = viaShow
-
--- | Computes the 'Arity' of a term.
+-- | Computes the 'Utils.Arity' of a term. Also returns the function body, for checking whether
+-- it's `Utils.acceptable`.
 computeArity ::
-    Term tyname name uni fun a
-    -> Arity
+    Term tyname name uni fun ann
+    -> (Arity, Term tyname name uni fun ann)
 computeArity = \case
-    LamAbs _ _ _ body -> MkTerm : computeArity body
-    TyAbs _ _ _ body  -> MkType : computeArity body
+    LamAbs _ _ _ body ->
+      let (nextArgs, nextBody) = computeArity body in (TermParam:nextArgs, nextBody)
+    TyAbs _ _ _ body  ->
+      let (nextArgs, nextBody) = computeArity body in (TypeParam:nextArgs, nextBody)
     -- Whenever we encounter a body that is not a lambda or type abstraction, we are done counting
-    _                 -> []
+    tm                -> ([],tm)
+
+-- | A term or type argument.
+data Arg tyname name uni fun ann =
+  MkTermArg (Term tyname name uni fun ann)
+  | MkTypeArg (Type tyname uni ann)
+
+-- | A list of type or term argument(s) that are being applied.
+type Args tyname name uni fun ann = [Arg tyname name uni fun ann]
+
+-- | A pair of argument and the annotation of the term being applied to,
+-- so a term that was traversed can be built back with `mkApps`.
+type ArgsWithAnns tyname name uni fun ann =
+  [(Arg tyname name uni fun ann, ann)]
+
+-- | Takes a term or type application and returns the function
+-- being applied and the arguments to which it is applied
+collectArgs :: Term tyname name uni fun ann
+  -> (Term tyname name uni fun ann, ArgsWithAnns tyname name uni fun ann)
+collectArgs tm
+  = go tm []
+  where
+    go (Apply ann f a) as      = go f ((MkTermArg a, ann):as)
+    go (TyInst ann f tyArg) as = go f ((MkTypeArg tyArg, ann):as)
+    go t as                    = (t, as)
+
+-- | Apply a list of term and type arguments to a function in potentially a nested fashion.
+mkApps :: Term tyname name uni fun ann
+  -> ArgsWithAnns tyname name uni fun ann
+  -> Term tyname name uni fun ann
+mkApps f ((MkTermArg tmArg, ann) : args) = mkApps (Apply ann f tmArg) args
+mkApps f ((MkTypeArg tyArg, ann) : args) = mkApps (TyInst ann f tyArg) args
+mkApps f []                              = f
+
+-- | Given the arity of a function, and the list of arguments applied to it, return whether it is
+-- fully applied or not.
+isFullyApplied :: Arity -> Args tyname name uni fun ann -> Bool
+isFullyApplied [] _ = True -- The function needs no more arguments
+isFullyApplied (_lam:_lams) [] = False -- under-application
+isFullyApplied (hdLams:tlLams) (hdArg:tlArg) =
+  case (hdLams, hdArg) of
+    (TermParam, MkTermArg _) -> isFullyApplied tlLams tlArg
+    (TypeParam, MkTypeArg _) -> isFullyApplied tlLams tlArg
+    _                        ->
```

Comment:
I asked the same here: https://github.com/input-output-hk/plutus/pull/5217#discussion_r1151772400

I think it's marginally better to leave things alone in this case: something bad is going on, and we can just decide to do nothing rather than try and make an argument about how we're not making things worse. But the comment should make that clear.